File: backend/visual_personal.py
"""Explicit transfers of student-authored visual work to personal annotations."""
import hashlib
from typing import Annotated, Literal

from fastapi import HTTPException
from pydantic import Field, StringConstraints
from sqlalchemy import text
from sqlalchemy.orm import Session

from . import models, schemas
from .message_diagrams import session_questionnaire
from .routes.learner_profile import _latest_revision
from .visual_tools import LABELS, StrictModel, load_workspace

NOTEBOOK_FIELDS = ('context', 'goal', 'main_difficulty', 'strengths', 'weaknesses', 'notes')
READING_FIELDS = ('note',)
READING_MAX_CHARS = 2000


class PersonalTransfer(StrictModel):
    revision: int = Field(ge=0)
    entry: str = Field(min_length=1, max_length=100)
    destination: Literal['notebook', 'reading']
    field: str = Field(min_length=1, max_length=40)
    expected_text: Annotated[str, StringConstraints(strip_whitespace=False)] = Field(default='', max_length=2000)
    text: str = Field(min_length=1, max_length=2000)
    language: str = Field(default='it', max_length=10)


def personal_context(db: Session, session_id: str, username: str, language: str = 'it') -> dict:
    from .routes.survey import INSTRUMENT_TYPES

    questionnaire = session_questionnaire(db, session_id)
    if not questionnaire:
        log = db.query(models.Log).filter_by(session_id=session_id, username=username, action='chat_message').order_by(models.Log.id.desc()).first()
        questionnaire = log.questionnaire_type if log else None
    if questionnaire not in INSTRUMENT_TYPES:
        questionnaire = None
    notebook = _latest_revision(db, username)
    labels = LABELS.get(language[:2], LABELS['en'])
    # «La mia lettura» appartiene alla compilazione di questa sessione: senza compilazione non c'è.
    result = db.query(models.QuestionnaireResult).filter_by(session_id=session_id, username=username).first()
    reading = db.query(models.ResultReading).filter_by(session_id=session_id, username=username).first() if result else None
    return {
        'questionnaire_type': questionnaire,
        'limits': {'notebook': schemas.LEARNER_PROFILE_MAX_FIELD_CHARS, 'reading': READING_MAX_CHARS},
        'sources': {kind: f'{labels[0]} · {labels[index]}' for kind, index in [('actions', 1), ('cards', 6), ('comparison', 11)]},
        'notebook': {key: (notebook.data or {}).get(key, '') if notebook else '' for key in NOTEBOOK_FIELDS},
        'reading': {'session_id': session_id, 'note': reading.note if reading else ''} if result else None,
    }


def transfer_to_personal(db: Session, session_id: str, username: str, update: PersonalTransfer) -> dict:
    if db.get_bind().dialect.name == 'postgresql':
        key = int.from_bytes(hashlib.sha256(f'visual-personal:{username}'.encode()).digest()[:8], 'big', signed=True)
        db.execute(text('SELECT pg_advisory_xact_lock(:key)'), {'key': key})
    current = load_workspace(db, session_id, username)
    if current['revision'] != update.revision:
        raise HTTPException(409, 'personal_conflict')
    kind, _, entry_id = update.entry.partition(':')
    workspace = current['workspace']
    entries = workspace.get(kind, []) if kind in ('actions', 'cards') else workspace['comparison']['options'] if kind == 'comparison' else []
    entry = next((item for item in entries if item['id'] == entry_id), None)
    if not entry:
        raise HTTPException(422, 'personal_invalid')
    fields = NOTEBOOK_FIELDS if update.destination == 'notebook' else READING_FIELDS
    if update.field not in fields:
        raise HTTPException(422, 'personal_invalid')
    context = personal_context(db, session_id, username, update.language)
    block = f"{update.text}\n({context['sources'][kind]})"
    if update.destination == 'notebook':
        row = _latest_revision(db, username)
        data = dict(row.data or {}) if row else {}
        limit = schemas.LEARNER_PROFILE_MAX_FIELD_CHARS
    else:
        if context['reading'] is None:
            raise HTTPException(422, 'personal_invalid')
        row = db.query(models.ResultReading).filter_by(session_id=session_id, username=username).with_for_update().first()
        data = {'note': row.note if row else ''}
        limit = READING_MAX_CHARS
    previous = str(data.get(update.field) or '')
    if block in previous:
        return {'status': 'duplicate', 'context': context}
    if previous != update.expected_text:
        raise HTTPException(409, 'personal_conflict')
    value = '\n\n'.join(part for part in (previous, block) if part)
    if len(value) > limit:
        raise HTTPException(422, 'personal_limit')
    data[update.field] = value
    if update.destination == 'notebook':
        row = models.LearnerProfileRevision(username=username, data=data, source='manual', session_id=session_id)
        db.add(row)
    elif row:
        row.note = value
    else:
        result = db.query(models.QuestionnaireResult).filter_by(session_id=session_id, username=username).first()
        db.add(models.ResultReading(username=username, session_id=session_id, questionnaire_type=result.questionnaire_type, note=value))
    db.commit()
    return {'status': 'saved', 'context': personal_context(db, session_id, username, update.language)}

File: frontend/src/lib/visual-personal.ts
// @ts-expect-error -- Node's direct TypeScript runner requires the extension.
import { emptyWorkspace, workspaceText, cardColumnsOf, type VisualWorkspace } from './visual-tools.ts';
// @ts-expect-error -- Node's direct TypeScript runner requires the extension.
import { readingText } from './i18n-reading.ts';

export const notebookFields: Record<string, string> = {
    context: 'lp.field.context', goal: 'lp.field.goal', main_difficulty: 'lp.field.difficulty',
    strengths: 'lp.field.strengths', weaknesses: 'lp.field.weaknesses', notes: 'lp.field.notes',
};
export type PersonalContext = {
    questionnaire_type: string | null;
    limits: { notebook: number; reading: number };
    sources: Record<string, string>;
    notebook: Record<string, string>;
    // «La mia lettura» della compilazione di questa sessione; null se la sessione non ha compilazione.
    reading: { session_id: string; note: string } | null;
};
export const readingLabel = (lang: string) => `${readingText(lang, 'title')} · ${readingText(lang, 'note')}`;
export type TransferEntry = { id: string; label: string; text: string; source: string };
export type ImportTarget = 'cards' | 'actions' | 'comparison';

export function visualEntries(work: VisualWorkspace, l: (key: string) => string): TransferEntry[] {
    const asText = (part: VisualWorkspace) => workspaceText(part, l).slice(l('handoff').length).trim();
    return [
        ...work.actions.map(item => ({ id: `actions:${item.id}`, label: item.title,
            text: asText({ ...emptyWorkspace(), actions: [item] }), source: 'actions' })),
        ...work.cards.map(item => ({ id: `cards:${item.id}`, label: item.text,
            text: asText({ ...emptyWorkspace(), cards: [item] }), source: 'cards' })),
        ...(work.comparison.options.length ? [{ id: `comparison:${work.comparison.options[0].id}`, label: l('comparison'),
            text: asText({ ...emptyWorkspace(), comparison: work.comparison }), source: 'comparison' }] : []),
    ];
}

export function annotationEntries(data: PersonalContext, t: (key: string) => string, l: (key: string) => string, lang: string): TransferEntry[] {
    const notebook = Object.entries(notebookFields).filter(([key]) => data.notebook[key]?.trim()).map(([key, label]) => ({
        id: `notebook_${key}`, label: `${l('notebook')} · ${t(label)}`, text: data.notebook[key], source: `${l('notebook')} · ${t(label)}`,
    }));
    const reading = data.reading?.note.trim() ? [{
        id: 'reading_note', label: `${readingLabel(lang)} · ${data.questionnaire_type ?? ''}`.trim(), text: data.reading.note,
        source: `${readingLabel(lang)} · ${data.questionnaire_type ?? ''}`.trim().slice(0, 300),
    }] : [];
    return [...notebook, ...reading];
}

export function importAnnotation(work: VisualWorkspace, entry: TransferEntry, target: ImportTarget, text: string, title: string): VisualWorkspace {
    const id = crypto.randomUUID();
    const items = target === 'comparison' ? work.comparison.options : work[target];
    if (items.some(item => item.source === entry.source && ('text' in item ? item.text === text.trim()
        : 'detail' in item ? item.detail === text.trim() && item.title === title.trim() : item.title === text.trim()))) throw new Error('personalDuplicate');
    if (items.length >= (target === 'comparison' ? 3 : 30)) throw new Error('limit');
    if (!text.trim() || text.length > (target === 'actions' ? 1000 : target === 'cards' ? 600 : 160)
        || (target === 'actions' && (!title.trim() || title.length > 160))) throw new Error('personalLength');
    const source = entry.source;
    if (target === 'cards') return { ...work, cards: [...work.cards, { id, text: text.trim(), bucket: cardColumnsOf(work)[0].id, source }] };
    if (target === 'actions') return { ...work, actions: [...work.actions, { id, title: title.trim(), detail: text.trim(), stage: 'todo', reflection: '', source }] };
    return { ...work, comparison: { ...work.comparison, options: [...work.comparison.options, { id, title: text.trim(), source }] } };
}
